Core/src/IfcGeometryConverter/FaceConverter.h

```diff
@@ -726,85 +726,89 @@ namespace OpenInfraPlatform {
 					throw oip::UnhandledException(surface);
 				}
 
-				/*! \brief  Converts a list of \c IfcFace -s to a polygon and adds it to the carve PolyhedronData vector.
+				/*! \brief  Converts a list of \c IfcFace -s to a polyhedron and adds it to the carve PolyhedronData vector.
 				\param[in]	faces		List \c IfcFace entity to be interpreted.
 				\param[in]	pos			The relative location of the origin of the representation's coordinate system within the geometric context.
-				\param[out]	itemData	Polygon carve polygon.
+				\param[out]	itemData	Polyhedron carve polyhedron.
 				\note The \c IfcFaceList can be an open or closed shell.
 				*/
 
 				void convertIfcFaceList(const std::vector<EXPRESSReference<typename IfcEntityTypesT::IfcFace>>& faces,
 					const carve::math::Matrix& pos,
 					std::shared_ptr<ItemData> itemData)  const noexcept(false) 
 				{
-					// Carve polygon of the converted face list
-					std::shared_ptr<carve::input::PolyhedronData> polygon(new carve::input::PolyhedronData());
+					// Carve polyhedron of the converted face list
+					std::shared_ptr<carve::input::PolyhedronData> polyhedron(new carve::input::PolyhedronData());
 
-					// Contains polygon indices of vertices (x,y,z converted to string)
-					std::map<std::string, uint32_t> polygonIndices;
+					// Contains polyhedron indices of vertices (x,y,z converted to string)
+					std::map<std::string, uint32_t> polyhedronIndices;
 
 					// Loop through all faces
 					for (const auto& face : faces) 
 					{
-						// get mesh data into polygon and polygonIndices
-						convertIfcFace(face, pos, polygon, polygonIndices);
+						// get mesh data into polyhedron and polyhedronIndices
+						convertIfcFace(face, pos, polyhedron, polyhedronIndices);
 					}
 
 					// IfcFaceList can be a closed or open shell, so let the calling function decide where to put it
-					itemData->open_or_closed_polyhedrons.push_back(polygon);
+					itemData->open_or_closed_polyhedrons.push_back(polyhedron);
 				}
 
-				/*! \brief  Converts \c IfcFace to a vector of face vertices.
-				\param	face \c IfcFace entity to be interpreted.
-				\param	pos
-				\param	polygon
-				\param	polygonIndices
+				/*! \brief  Converts \c IfcFace or sub-entity to a vector of face vertices, respectively the mesh of a polyhedron.
+				\param[in]		face				\c IfcFace entity to be converted.
+				\param[in]		pos					The relative location of the origin of the representation's coordinate system within the geometric context.
+				\param[in,out]	polyhedron			Carve polyhedron of the converted face.
+				\param[in,out]	polyhedronIndices	Contains polyhedron indices of vertices (x,y,z converted to string).

```

Comment:
At the moment, there is no explicit function which converts the `polyhedronIndices`. However, thanks for your comment - it showed me the need of refactoring at this point. For another programmer, it isn't easy to see from where the variable `polyhedronIndices` comes from, how it has to be initialised, does it need some kine of default pre-set, and so on.

**A few words about how it works right now**
The variable `std::map<std::string, uint32_t> polyhedronIndices` is declared at one place - in this case inside the function `convertIfcFaceList`. From there it's passed by reference to all functions which work with it.

'Work' refers to the adding a vertex to `polyhedron`. In that case, a `stringstream` with the vertex coordinates is formed and searched inside of `polyhedronIndices`. If the `stringstream` exists, the stored vertex index is used. Otherwise, the vertex is added to `polyhedron` and the `stringstream` is added to `polyhedronIndices`. This task is done by the following sample code:

```cpp
// set string id and search for existing vertex in polyhedron
std::stringstream vertexString;
vertexString << vertex3D.x << " " << vertex3D.y << " " << vertex3D.z;
auto itFound = polyhedronIndices.find(vertexString.str());

uint32_t index = 0;
if (itFound != polyhedronIndices.end()) {
	index = itFound->second;
}
else {
	index = polyhedron->addVertex(vertex3D);
	polyhedronIndices[vertexString.str()] = index;
}

// use index (depends on context)
```

The variable polyhedronIndices is passed by reference through the multiple function calls because its content is filled incrementally and reused in the next calls. In this way, it's responsible for the topological merge of coincident vertices.

The above sample code can be found at multiple places throughout the project - sometimes with little adaptions:
- Subfunctions of `FaceConverterT::convertIfcFaceList`:
    - `FaceConverterT::addTriangleToPolyhedronData`
    - `FaceConverterT::triangulateFace`
    - `ItemData::mergePolyhedronsIntoOnePolyhedron`
- Local use in `FaceConverterT::convertIfcBSplineSurface`
- Local use in `ProfilConverterT::SearchExistingVertex`
- Local use in `SolidModelConverterT::convertIfcSpecificSolidModel` (function is commented out)

**Thoughts about how to refactor**
Because the same task is done at multiple places in the project, it should be reduced into one central function. Within this change, the variable `polyhedronIndices` should be somehow bound to the corresponding polyhedron, hence another programmer don't has to care that much about this variable.

As one conclusion, the information in `polyhedronIndices` is just another representation of the content from the member variable `std::vector<carve::geom3d::Vector> points` of the class `carve::input::PolyhedronData`, respectively `carve::input::PolylineSetData` (`points` is inherited from the class `carve::input::VertexData` where it's declared). Probably, the type `std::map<std::string, uint32_t>` was choosen to have access the `find` function of `std::map`.

To be accurate, `polyhedronIndices` should be a member variable in `VertexData` and the above samle code becomes a memberfunction inside there. However, this is outside of our actual namespace.

Inside the namespace OIP, the class `ItemData` would be the next alternative for the member variable and its depending functions - the function `mergePolyhedronsIntoOnePolyhedron` which works within this topic of topological merging is already there. The first idea was to keep and use the existing instances of `PolyhedronData` and `PolylineSetData`. One disadvantage of this was, that we would need one `polyhedronIndices`-variable for each instance of `PolyhedronData` and `PolylineSetData`, thus four vectors of them (for `closed_polyhedrons`, `open_polyhedrons`, `open_or_closed_polyhedrons` and `polylines`). Futhermore, assuming on function is in execution on one instance of a polyhedron, it will be quite difficult to determine the corresponding `polyhedronIndices`-variable.

Reply:
With reference to todays meeting, the described problem will be subject of a new github-issue.
However, I want to update the doxygen comment to make more clear where the parameter `polyhedronIndices` comes from and how to handle it.

Reply:
Doxygen comments are updated by commit ac64eb1; the result is shown in the comment [below](https://github.com/tumcms/Open-Infra-Platform/pull/462#issuecomment-1046816351).